### PR TITLE
cornered box annotator

### PR DESCRIPTION
**Pull request by hardikdava**
# Description

As discussed in #168  and #170 . Merging new annotation format for initial commit.

@SkalskiP it might needed to refine a bit.

## Type of change

-   [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Comment by hardikdava** [comment]
@SkalskiP Please merge this branch before further work. I already finished some implementation.

**Comment by hardikdava** [comment]
@SkalskiP #49 is also included in this PR.

**Comment by hardikdava** [comment]
@SkalskiP I changed code style to consistent style as `Annotators.annotate(scene, detection)` format. Please review it and merge the PR.

**Comment by hardikdava** [comment]
@SkalskiP `EllispseAnotator` is also added to the `Anotator` API. #16

**Review by SkalskiP — APPROVED** [review]

File: supervision/detection/track.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        self.position = position
+        self.max_length = max_length
+        self.frame_counter = 0
+        self.storage = np.zeros((0, 5))
+
+    def update(self, frame_counter: int, detections: Detections) -> None:
+        if detections.xyxy.shape[0] > 0:
+            xyxy = detections.xyxy
+
+            if self.position == Position.CENTER:
```
Comment (SkalskiP) [inline supervision/detection/track.py:43]: That part of the logic is already implemented in `sv.Detections.get_anchor_coordinates`. Let's refactor our logic here to use the `get_anchor_coordinates` method. Also:

- Let's expand `Position` with:
    -  `TOP_CENTER = "TOP_CENTER"`
    -  `TOP_LEFT = "TOP_LEFT"`
    -  `TOP_RIGHT = "TOP_RIGHT"`
    -  `CENTER_LEFT = "CENTER_LEFT"`
    -  `CENTER_RIGHT = "CENTER_RIGHT"`
    -  `BOTTOM_LEFT = "BOTTOM_LEFT"`
    -  `BOTTOM_RIGHT = "BOTTOM_RIGHT"`

- Let's expand the `get_anchor_coordinates` implementation to handle new `Position`'s
- We need to add unit tests for `get_anchor_coordinates`, but I can handle that. Just let me know when you'll be ready.

File: supervision/detection/track.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            >>> for frame in sv.get_video_frames_generator(source_path='source_video.mp4'):
+            >>>     detections = sv.Detections(...)
+            >>>     tracked_objects = tracker(...)
+            >>>     tracked_detections = sv.Detections(tracked_objects)
+            >>>     track_storage.update(tracked_detections)
+        """
+        self.position = position
+        self.max_length = max_length
+        self.frame_counter = 0
+        self.storage = np.zeros((0, 5))
```
Comment (SkalskiP) [inline supervision/detection/track.py:37]: Curious why you decided to use `np.ndarray` as a storage engine.

**Review by SkalskiP — CHANGES_REQUESTED** [review]

**Comment by SkalskiP** [comment]
@hardikdava I left you few comments already. I'm taking a look at the rest of the code, as well as testing it. But there are a few things that I need to think about tomorrow. Please start implementing the changes I already listed. 🙏🏻

File: supervision/detection/track.py (hunk 1 above)
Comment (hardikdava) [reply supervision/detection/track.py]: @SkalskiP , I made changes and added extra anchor points.

File: supervision/detection/track.py (hunk 2 above)
Comment (hardikdava) [reply supervision/detection/track.py:37]: > Curious why you decided to use `np.ndarray` as a storage engine.

Hi @SkalskiP, I am trying to rewrite `TrackerStorage` in the format of `Detections`. I will let you know once I am finished.

Comment (hardikdava) [reply supervision/detection/track.py]: @SkalskiP I am done with consistent format for `TrackerStorage`. You can review it now.

File: supervision/annotators/core.py [hunk 3]
```diff
@@ -74,20 +173,604 @@ def annotate(self, scene: np.ndarray, detections: Detections) -> np.ndarray:
...
-        pass
+    def __init__(
+        self,
+        color: Union[Color, ColorPalette] = ColorPalette.default(),
+        thickness: int = 2,
+        text_color: Color = Color.black(),
+        text_scale: float = 0.5,
+        text_thickness: int = 1,
+        text_padding: int = 10,
+        color_by_track: bool = False,
```
Comment (SkalskiP) [inline supervision/annotators/core.py:191]: Let's add `label_formatter: Callable[[Detections], List[str]]` argument.

The default `label_formatter` could be something like this.

```python
def default_label_formatter(detections: Detections) -> List[str]:
    return [
        str(class_id) 
        for class_id 
        in detections.class_id
    ]
```

If you want to use class mapping in your `label_formatter` you could do it like this:

```python
def build_label_formatter(classes: List[str]) -> Callable[[Detections], List[str]]:
    def default_label_formatter(detections: Detections) -> List[str]:
        return [
            classes[class_id] 
            for class_id 
            in detections.class_id
        ]
    return default_label_formatter
```

**Review by SkalskiP — CHANGES_REQUESTED** [review]

File: supervision/annotators/core.py (hunk 3 above)
Comment (hardikdava) [reply supervision/annotators/core.py:191]: Hi @SkalskiP :wave: , Thanks for the idea. I think it is too much for just to create `label`. This type of callback is not a good idea. It makes complicated everything. Instead, what do you think of this:

```python

class LabelAnnotator(BaseAnnotator):
	def __init__(..., labelmaps: Optionl[Dict[int, str]]=None):
		self.labelmaps = labelmap
		
	def annotate(self,  scene: np.ndarray,  detections: Detections):
	    for det in detections:
			if self.labelmaps:
		       	label=self.labelmaps[classid]
		    if not self.labelmaps:
			    labels = f"{classid}"				  

```
An example of labelmaps would be like:

```
labelmaps = {0: 'person', 1: '...', 2: '...'}
```
My concern is that if user has to write callback function then mostly they are not going to use this from `supervision` but instead they will use their own.

Comment (SkalskiP) [reply supervision/annotators/core.py:191]: Hi @hardikdava 👋🏻, label map is, of course, a lot simpler, but a mall problem I see is not flexible :/ It doesn't allow you to build complicated labels.

How about adding both `classes: Optional[List[str]], label_formatter: Optional[Callable[[Detections], List[str]]]`?

Comment (hardikdava) [reply supervision/annotators/core.py:191]: @SkalskiP, That seems fine to me. User can have flexibility more when they want to use some other format.

Comment (SkalskiP) [reply supervision/annotators/core.py:191]: @hardikdava exactly! There will likely be multiple people asking about label customization. And I want to support them as well :)

Comment (hardikdava) [reply supervision/annotators/core.py:203]: It is done. Ready for test.

Comment (SkalskiP) [reply supervision/annotators/core.py:203]: Awesome! 🔥 Let me test it. We will probably need to add unit tests for `get_anchor_coordinates`. Would you like me to take care of that?

Comment (hardikdava) [reply supervision/annotators/core.py:203]: Okay. I will try. What is command line for running tests? basic pytest or something special?

**Comment by hardikdava** [comment]
@SkalskiP I have added tests for `get_anchor_coordinates`

**Pull request merged**